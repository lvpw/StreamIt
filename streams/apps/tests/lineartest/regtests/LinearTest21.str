/** 
 * Test out combining pipelines of FIR filters (eg not replacing them with 
 * their combination when the cost of doing so is 
 * higher.
 **/
void->void pipeline LinearTest21 {
   add FloatSource();
   add FIRPipelineBad();
   add FIRPipelineGood();
   add FloatSink();
 }

 void->float stateful filter FloatSource {
   float x;
   init {
     x = 0;
   }
   work push 1 {
     push(x);
     x = x + 1;
   }
 }

 float->void filter FloatSink {
   work pop 1 {
     println(pop());
   }
 }

/** A linear pipeline that incurs more calculation if combined. **/
float->float pipeline FIRPipelineBad {
   add LinearFilterOne();
   add LinearFilterTwo();
 }

/** A linear pipeline that incurs less calculation if combined. **/
float->float pipeline FIRPipelineGood {
  add LinearFilterTwo();
  add LinearFilterOne();
 }

 float->float filter LinearFilterOne {
   work push 1 pop 1 peek 4 {
     push(4*peek(0) +
	  3*peek(1) +
	  2*peek(2) +
	  1*peek(3));
     pop();
   }
 }


 float->float filter LinearFilterTwo {
   work push 4 pop 1 peek 1 {
     float t = pop();
     push(8*t);
     push(7*t);
     push(6*t);
     push(5*t);
   }
 }
