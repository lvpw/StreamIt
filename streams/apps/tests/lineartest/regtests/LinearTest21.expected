FIRPipelineBad_1.matrix
[[5.0+0.0i 6.0+0.0i 7.0+0.0i 8.0+0.0i]
 [10.0+0.0i 12.0+0.0i 14.0+0.0i 16.0+0.0i]
 [15.0+0.0i 18.0+0.0i 21.0+0.0i 24.0+0.0i]
 [20.0+0.0i 24.0+0.0i 28.0+0.0i 32.0+0.0i]]
FIRPipelineBad_1.vector
[[0.0+0.0i 0.0+0.0i 0.0+0.0i 0.0+0.0i]]
FIRPipelineGood_1.matrix
[[70.0+0.0i]]
FIRPipelineGood_1.vector
[[0.0+0.0i]]
LinearFilterOne_1.matrix
[[1.0+0.0i]
 [2.0+0.0i]
 [3.0+0.0i]
 [4.0+0.0i]]
LinearFilterOne_1.vector
[[0.0+0.0i]]
LinearFilterOne_2.matrix
[[1.0+0.0i]
 [2.0+0.0i]
 [3.0+0.0i]
 [4.0+0.0i]]
LinearFilterOne_2.vector
[[0.0+0.0i]]
LinearFilterTwo_1.matrix
[[5.0+0.0i 6.0+0.0i 7.0+0.0i 8.0+0.0i]]
LinearFilterTwo_1.vector
[[0.0+0.0i 0.0+0.0i 0.0+0.0i 0.0+0.0i]]
LinearFilterTwo_2.matrix
[[5.0+0.0i 6.0+0.0i 7.0+0.0i 8.0+0.0i]]
LinearFilterTwo_2.vector
[[0.0+0.0i 0.0+0.0i 0.0+0.0i 0.0+0.0i]]
LinearWrapper_1.matrix
[[350.0+0.0i 420.0+0.0i 490.0+0.0i 560.0+0.0i]
 [700.0+0.0i 840.0+0.0i 980.0+0.0i 1120.0+0.0i]
 [1050.0+0.0i 1260.0+0.0i 1470.0+0.0i 1680.0+0.0i]
 [1400.0+0.0i 1680.0+0.0i 1960.0+0.0i 2240.0+0.0i]]
LinearWrapper_1.vector
[[0.0+0.0i 0.0+0.0i 0.0+0.0i 0.0+0.0i]]
